drivers/soundwire/cadence_master.c

```diff
@@ -511,11 +510,43 @@ cdns_fill_msg_resp(struct sdw_cdns *cdns,
 	return SDW_CMD_OK;
 }
 
+static int wait_for_tx_complete(struct sdw_cdns *cdns, unsigned int timeout)
+{
+	u32 int_status;
+	int i;
+
+	for (i = 0; i < timeout; i++) {
+		int_status = cdns_readl(cdns, CDNS_MCP_INTSTAT);
+		if (int_status & CDNS_MCP_INT_RX_WL) {
+			cdns_writel(cdns, CDNS_MCP_INTSTAT, CDNS_MCP_INT_RX_WL);
+			return 0;
+		}
+
+		msleep(1);
+	}
+	return -ETIME;
+}
+
+static void cdns_read_response(struct sdw_cdns *cdns)
+{
+	u32 num_resp, cmd_base;
+	int i;
+
+	num_resp = cdns_readl(cdns, CDNS_MCP_FIFOSTAT);
+	num_resp &= CDNS_MCP_RX_FIFO_AVAIL;
+
+	cmd_base = CDNS_MCP_CMD_BASE;
+
+	for (i = 0; i < num_resp; i++) {
+		cdns->response_buf[i] = cdns_readl(cdns, cmd_base);
+		cmd_base += CDNS_MCP_CMD_WORD_LEN;
+	}
+}
+
```

Comment:
A command in SoundWire is scheduled at every frame. that's 10us of waiting. In addition, the commands are serialized so the bus may be dealing with commands for another slave.
It'd be really bad to poll in this way in the *shared* interrupt thread, it will impact IPC and other things that we deal with in interrupt threads, we really need to use an asynchronous design.

Reply:
> A command in SoundWire is scheduled at every frame. that's 10us of waiting. 

We can shorten the waiting interval to less than 10us then.

> In addition, the commands are serialized so the bus may be dealing with commands for another slave.

I am a little bit lost here. This is also true when we wait interrupt to read response, right?

> It'd be really bad to poll in this way in the _shared_ interrupt thread, it will impact IPC and other things that we deal with in interrupt threads, we really need to use an asynchronous design.

Yes, it is indeed a problem. Let's close this PR and try to explain the needed of using a thread inside irq thread to Vinod.

Reply:
> > A command in SoundWire is scheduled at every frame. that's 10us of waiting.
> 
> We can shorten the waiting interval to less than 10us then.

No you cannot. this is linked to the bus frame rate, i.e. 48kHz. We have to wait for the command to be issued and get an answer from the slave, which triggers an interrupt when the RX FIFO fills.

> > In addition, the commands are serialized so the bus may be dealing with commands for another slave.
> 
> I am a little bit lost here. This is also true when we wait interrupt to read response, right?

The interrupt is initiated on the host side when the FIFO fills, but the FIFO will only fill when the master IP gets the response from the Slave device.

> > It'd be really bad to poll in this way in the _shared_ interrupt thread, it will impact IPC and other things that we deal with in interrupt threads, we really need to use an asynchronous design.
> 
> Yes, it is indeed a problem. Let's close this PR and try to explain the needed of using a thread inside irq thread to Vinod.